From: David S. Miller <davem@davemloft.net>
Subject: ipv6: Check dest prefix length on original route not copied one in rt6_alloc_cow().
Patch-mainline: 3.2
Git-commit: bb3c36863e8001fc21a88bebfdead4da4c23e848
References: bnc#757202
Acked-by: Jiri Bohac <jbohac@suse.cz>

After commit 8e2ec639173f325977818c45011ee176ef2b11f6 ("ipv6: don't
use inetpeer to store metrics for routes.") the test in rt6_alloc_cow()
for setting the ANYCAST flag is now wrong.

'rt' will always now have a plen of 128, because it is set explicitly
to 128 by ip6_rt_copy.

So to restore the semantics of the test, check the destination prefix
length of 'ort'.

Signed-off-by: David S. Miller <davem@davemloft.net>

Index: expanded/net/ipv6/route.c
===================================================================
--- expanded.orig/net/ipv6/route.c	2012-04-14 00:54:10.000000000 +0200
+++ expanded/net/ipv6/route.c	2012-04-14 00:55:04.000000000 +0200
@@ -710,7 +710,7 @@ static struct rt6_info *rt6_alloc_cow(co
 		int attempts = !in_softirq();
 
 		if (!(rt->rt6i_flags&RTF_GATEWAY)) {
-			if (rt->rt6i_dst.plen != 128 &&
+			if (ort->rt6i_dst.plen != 128 &&
 			    ipv6_addr_equal(&ort->rt6i_dst.addr, daddr))
 				rt->rt6i_flags |= RTF_ANYCAST;
 			ipv6_addr_copy(&rt->rt6i_gateway, daddr);
